tests/core/CoreTest.py

```diff
@@ -57,12 +57,278 @@ def generate_tasks(self):
+        self.assertEqual(len(dependency_tracker.get_dependencies(bear_d)), 1)
+
+        # Test path BearE -> BearD -> BearC.
+        self.assertEqual(len(dependency_tracker.get_dependencies(bear_d)), 1)
+
+        bear_c = dependency_tracker.get_dependencies(bear_d).pop()
+
+        self.assertIs(bear_c.section, section)
+        self.assertIs(bear_a.file_dict, filedict)
+        self.assertIsInstance(bear_c, BearC_NeedsB)
+
+        # Test path BearE -> BearD -> BearC -> BearB.
+        self.assertEqual(len(dependency_tracker.get_dependencies(bear_c)), 1)
+
+        bear_b = dependency_tracker.get_dependencies(bear_c).pop()
+
+        self.assertIs(bear_b.section, section)
+        self.assertIs(bear_a.file_dict, filedict)
+        self.assertIsInstance(bear_b, BearB)
+
+        # No more dependencies after BearB.
+        self.assertEqual(dependency_tracker.get_dependencies(bear_b), set())
+
+        # Finally check the bears_to_schedule
+        self.assertEqual(bears_to_schedule, {bear_a, bear_b})
+
+    def test_simple(self):
+        # Test simple case without dependencies.
+        section = Section('test-section')
+        filedict = {}
+
+        bear_a = BearA(section, filedict)
+        bear_b = BearB(section, filedict)
+
+        dependency_tracker, bears_to_schedule = initialize_dependencies(
+            {bear_a, bear_b})
+
+        self.assertTrue(dependency_tracker.all_dependencies_resolved)
+        self.assertEqual(bears_to_schedule, {bear_a, bear_b})
+
+    def test_reuse_instantiated_dependencies(self):
+        # Test whether pre-instantiated dependency bears are correctly
+        # (re)used.
+        section = Section('test-section')
+        filedict = {}
+
+        bear_b = BearB(section, filedict)
+        bear_c = BearC_NeedsB(section, filedict)
+
+        dependency_tracker, bears_to_schedule = initialize_dependencies(
+            {bear_b, bear_c})
+
+        self.assertEqual(dependency_tracker.dependants, {bear_c})
+        self.assertEqual(dependency_tracker.get_dependencies(bear_c), {bear_b})
+
+        self.assertEqual(bears_to_schedule, {bear_b})
+
+    def test_no_reuse_of_different_section_dependency(self):
+        # Test whether pre-instantiated bears which belong to different
+        # sections are not (re)used, as the sections are different.
+        section1 = Section('test-section1')
+        section2 = Section('test-section2')
+        filedict = {}
+
+        bear_b = BearB(section1, filedict)
+        bear_c = BearC_NeedsB(section2, filedict)
+
+        dependency_tracker, bears_to_schedule = initialize_dependencies(
+            {bear_b, bear_c})
+
+        self.assertEqual(dependency_tracker.dependants, {bear_c})
+        dependencies = dependency_tracker.dependencies
+        self.assertEqual(len(dependencies), 1)
+        dependency = dependencies.pop()
+        self.assertIsInstance(dependency, BearB)
+        self.assertIsNot(dependency, bear_b)
```

Comment:
the idea is that if the section is different, a new bear get instantiated. Imo this is okay^^